### PR TITLE
Toric covered schemes 

### PR DESCRIPTION
**Pull request by HechtiDerLachs**
This is a remake of my previous PR to bring together the toric varieties by @HereAround and @lkastner  with our schemes package.

### CODE REVIEW

**Comment by HereAround** [comment]
Proposal: Right now, the code seems to lead to an error when executed for P2 x P2 (explicitly, one gluing that should be there cannot be computed right now, because the cones in question only intersect in the origin). Once we figured this out, it would be great to have this as example/test.

**Comment by HereAround** [comment]
Another thought: Could we try to marry the coordinate names of the affine patches with (fractions of) the homogeneous coordinates in the cox ring?

**Comment by simonbrandhorst** [comment]
How about introducing toric morphisms? If you have two cones $C \subseteq D$, I seem to remember that there is a morphism 
of the corresponding toric varieties. That would be very convenient to build toric varieties from a fan and I guess also to
construct  toric morphisms on toric covered schemes. I am thinking of toric blowups.

**Comment by HereAround** [comment]
Yet another thought: Closed subvarieties in toric varieties have basic support. It would be cool to turn them into schemes as well (typically they are not toric, so any test for smoothness then relies on schemes/ring techniques).

**Comment by HereAround** [comment]
> How about introducing toric morphisms? If you have two cones C⊆D, I seem to remember that there is a morphism of the corresponding toric varieties. That would be very convenient to build toric varieties from a fan and I guess also to construct toric morphisms on toric covered schemes. I am thinking of toric blowups.

Very good thought! Aside: https://github.com/oscar-system/Oscar.jl/pull/1460 is a first draft towards toric morphisms.

File: experimental/Schemes/Types.jl [hunk 1]
```diff
@@ -985,3 +987,138 @@ mutable struct VarietyFunctionFieldElem{FracType<:AbstractAlgebra.Generic.Frac,
...
+
+        G = SimpleGlueing(X, Y, f, g)
+        add_glueing!(cov, G)
+      end
+    end
+
+    # TODO: Investigate whether or not we can make use of the Hasse Diagram 
+    # to take shortcuts in filling up the remaining glueings. 
+    # We conjecture, that the composition of the already existing glueings 
+    # is already maximal and so a lot of lengthy computations can be skipped.
```
Comment (HechtiDerLachs) [inline experimental/Schemes/Types.jl]: @lkastner If I remember correctly, you were looking into how one could make use of the Hasse diagram for chasing and extracting glueings, right? Let us meet again one of these days and try to work this stuff out.

File: experimental/Schemes/Types.jl [hunk 2]
```diff
@@ -985,3 +987,138 @@ mutable struct VarietyFunctionFieldElem{FracType<:AbstractAlgebra.Generic.Frac,
...
+    fill_transitions!(cov)
```
Comment (HechtiDerLachs) [inline experimental/Schemes/Types.jl]: For now, this command seems to take care of things. But it is doing a lot of tedious background work involving multiple Groebner basis computations. So we should eventually replace it. 

But yes: It works for now!

File: test/Schemes/ToricSchemes.jl [hunk 3]
```diff
@@ -0,0 +1,10 @@
+@testset "toric schemes" begin
+  S = hirzebruch_surface(3)
+  X = ToricCoveredScheme(S)
+  @test issmooth(X)
+
+  IP1 = projective_space(NormalToricVariety, 1)
+  IP1xIP1 = IP1*IP1
+  Y = ToricCoveredScheme(IP1xIP1)
+  @test length(values(glueings(default_covering(Y)))) == 16
```
Comment (HechtiDerLachs) [inline test/Schemes/ToricSchemes.jl]: See this example: What was causing the trouble before is now running. Looking forward to more interesting examples in the future.

**Comment by HereAround** [comment]
@HechtiDerLachs Could you please rebase to the current master?

**Comment by HereAround** [comment]
@HechtiDerLachs and I discussed this PR both yesterday and today (February 15 and 16). Based on this, we decided to propose a version that hopefully could be merged quickly (latest during the weekend - see below for the rationals).

Please notice, that this first version would merely become part of the experimental section of OSCAR. This is because the code depends on the covered schemes, which are still in their experimental status. Personally, I would propose to consider a migration of the covered schemes into the official OSCAR part once @HechtiDerLachs is back from his vacation. At least, based on my discussions with @HechtiDerLachs during this week, the code would seem to be sufficiently stable. Once complete, the toric covered schemes could be migrated into the official OSCAR code too.

Please notice that the toric covered schemes are a natural point to connect with https://github.com/Julia-meets-String-Theory/FTheoryTools.jl (cf. https://github.com/Julia-meets-String-Theory/FTheoryTools.jl/pull/26). This is part of the motivation why @HechtiDerLachs and I decided to push the matter.

After reasons to consider a timely merge are as follows:
1. Next week, the big renaming takes place. So merging before or during the upcoming weekend would mean that this code can benefit from the upcoming big changes in the code sprint next week.
2. There is a chance that @fingolfin wants to make a new OSCAR release before the big renaming. (After all, the big renaming hast the potential to break OSCAR for a while. So having a working version just before this adventurous move seems rather reasonable). If so, then the toric covered schemes would be available to the outside world by the beginning of the next week. With an eye towards `FTheoryTools`, this could be beneficial.

Of course, we have to wait for the tests to pass. I am confident they will. So with that said, @lkastner, @simonbrandhorst @afkafkafk13 what are your thoughts on this PR?

File: docs/src/Experimental/ToricSchemes/intro.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+## Long term goals
+
+We aim for a robust interface between toric varieties and schemes.
+
+
+## Contact
+
+Please direct questions about this part of OSCAR to the following people:
+* [Martin Bies](https://martinbies.github.io/),
+* [Matthias Zach](https://www.mathematik.uni-kl.de/en/agag/people/members/seite).
```
Comment (afkafkafk13) [inline docs/src/Experimental/ToricSchemes/intro.md:43]: .../members/**seite** does not seem to be the right page.

File: experimental/Schemes/ToricSchemes/NormalToricSchemes/Type.jl [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        # Now glue X and Y
+        
+        # Step 1: Find the cone of the intersection of X and Y.
+        facet = intersect(CX, CY)
+        # harvest the dual cone
+        dual_facet = polarize(facet)
+        if !(dim(facet) == dim(CX) - 1)
+          continue
+        end
```
Comment (afkafkafk13) [inline experimental/Schemes/ToricSchemes/NormalToricSchemes/Type.jl:34]: Is there a reason, why you polarize before testing (without using of dual_facet in the test) whether you want to pursue this case or not?

File: experimental/Schemes/ToricSchemes/NormalToricSchemes/Type.jl [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        set_attribute!(f, :inverse, g)
+        
+        G = SimpleGlueing(X, Y, f, g)
+        add_glueing!(cov, G)
+      end
+    end
+    
+    # TODO: Improve the gluing (lazy gluing) or try to use the Hasse diagram.
+    # TOOD: For now, we conjecture, that the composition of the computed glueings is sufficient to deduce all glueings.
+    fill_transitions!(cov)
```
Comment (simonbrandhorst) [inline experimental/Schemes/ToricSchemes/NormalToricSchemes/Type.jl:106]: Quick and dirty? Will this result in glueing over `SpecOpens`  which we know to be difficult?
Which patches are missing? I have the impression that you have glued all `n(n+1)/2` pairs in `patch_list` the toric way already?

**Review by simonbrandhorst — APPROVED** [review]
Looks good to me. Thanks for pushing this.

**Review by afkafkafk13 — COMMENTED** [review]
Looks good. Most of the meat is in NormalToricSchemes/Types.jl, the rest is mostly documentation and one-liners to introduce commands.

2 small comments at particular pieces of code

File: experimental/Schemes/ToricSchemes/NormalToricSchemes/Type.jl (hunk 6 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/ToricSchemes/NormalToricSchemes/Type.jl:106]: This needs refactoring, indeed. But we don't have the time for this right now. We don't expect `SpecOpen`s to be used, due to a special catch-block, but the computations are nevertheless lengthy. At the beginning, we only compute the glueings which correspond to codimension 1 intersections of maximal cones in the fan (see some if-continue-block further up). So filling is non-trivial here.

File: docs/src/Experimental/ToricSchemes/intro.md (hunk 4 above)
Comment (HechtiDerLachs) [reply docs/src/Experimental/ToricSchemes/intro.md:43]: Well, but it works, doesn't it? I don't know if the RPTU already provides a website for me and if yes, where to find it.

File: experimental/Schemes/ToricSchemes/NormalToricSchemes/Type.jl (hunk 5 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/ToricSchemes/NormalToricSchemes/Type.jl:34]: Probably yes. Frankly, I don't remember all details from the discussion with @lkastner from half a year ago.

**Comment by HereAround** [comment]
Thank you for the quick merge. (I was side-tracked, and just made it back to this PR... Sorry for the late response).

Long story short: As mentioned above, we definitely want to take a closer look at the glueings in the future. For now, this is just a first version, that is hopefully (-> experimental) working.

I shall try to investigate more in the upcoming weeks and discuss this at length once @HechtiDerLachs is back from his cool vacation. Also, @lkastner might have a few ideas.

**Pull request merged**